### PR TITLE
reduce realloc overhead, MacOS compile fix

### PR DESCRIPTION
**Pull request by ViperCraft**

### CODE REVIEW

File: src/annoylib.h [hunk 1]
```diff
@@ -827,21 +828,30 @@ template<typename S, typename T, typename Distance, typename Random>
...
+    typedef std::vector<qpair_t> qvector_t;
+
+    size_t const roots_size = _roots.size();
 
     if (search_k == (size_t)-1)
-      search_k = n * _roots.size(); // slightly arbitrary default value
+      search_k = n * roots_size; // slightly arbitrary default value
+
+    qvector_t qvector;
+    qvector.reserve(roots_size); // prealloc queue
```
Comment (erikbern) [inline src/annoylib.h]: could probably preallocate search_k items (not super important)

File: src/annoylib.h [hunk 2]
```diff
@@ -855,29 +865,32 @@ template<typename S, typename T, typename Distance, typename Random>
...
 
     size_t m = nns_dist.size();
-    size_t p = n < m ? n : m; // Return this many items
-    std::partial_sort(nns_dist.begin(), nns_dist.begin() + p, nns_dist.end());
-    for (size_t i = 0; i < p; i++) {
+    if( n < m ) // Has more than N results, so get only top N
+      std::partial_sort(nns_dist.begin(), nns_dist.begin() + n, nns_dist.end());
+    else
+      std::sort(nns_dist.begin(), nns_dist.end());
+    for (size_t i = 0, p = std::min(p, n); i < p; i++) {
```
Comment (erikbern) [inline src/annoylib.h]: where is `p` defined? i'm not sure if this code is a simplification

File: setup.py [hunk 3]
```diff
@@ -47,7 +47,7 @@
     cputune = ['-mcpu=native',]
 
 if os.name != 'nt':
-    compile_args = ['-O3', '-ffast-math', '-fno-associative-math']
+    compile_args = ['-O3', '-ffast-math', '-fno-associative-math', '-std=c++11']
```
Comment (erikbern) [inline setup.py]: is there a reason we need to rely on c++11? i'd rather drop that if possible

File: src/annoylib.h [hunk 4]
```diff
@@ -48,6 +48,7 @@ typedef signed __int32    int32_t;
 #include <algorithm>
 #include <queue>
 #include <limits>
+#include <utility>
```
Comment (erikbern) [inline src/annoylib.h]: what's this needed for?

File: examples/precision_test.cpp [hunk 5]
```diff
@@ -14,7 +14,7 @@
 #include <chrono>
 #include <algorithm>
 #include <map>
-
+#include <random>
```
Comment (erikbern) [inline examples/precision_test.cpp]: what's this needed for?

**Comment by erikbern** [comment]
I like the addition of const methods.

Does the other code actually solve some OS X compilation issue?

File: src/annoylib.h (hunk 4 above)
Comment (ViperCraft) [reply src/annoylib.h]: hmm, sorry, from another experiment, I'll remove this

File: examples/precision_test.cpp (hunk 5 above)
Comment (ViperCraft) [reply examples/precision_test.cpp]: not compiled with clang on MacOS:
```c++
compiling precision example...
precision_test.cpp:22:29: error: use of undeclared identifier 'generator'; did you mean 'operator'?
        std::default_random_engine generator;
                                   ^~~~~~~~~
                                   operator
precision_test.cpp:22:7: error: no type named 'default_random_engine' in namespace 'std'
        std::default_random_engine generator;
        ~~~~~^
precision_test.cpp:22:38: error: expected a type
        std::default_random_engine generator;
                                            ^
precision_test.cpp:23:7: error: no member named 'normal_distribution' in namespace 'std'
        std::normal_distribution<double> distribution(0.0, 1.0);
        ~~~~~^
precision_test.cpp:23:33: error: expected '(' for function-style cast or type construction
        std::normal_distribution<double> distribution(0.0, 1.0);
                                 ~~~~~~^
precision_test.cpp:23:35: error: use of undeclared identifier 'distribution'
        std::normal_distribution<double> distribution(0.0, 1.0);
                                         ^
precision_test.cpp:38:27: error: use of undeclared identifier 'generator'
                        vec[z] = (distribution(generator));

File: src/annoylib.h (hunk 1 above)
Comment (ViperCraft) [reply src/annoylib.h]: fixed

**Comment by ViperCraft** [comment]
no, generally this is just realloc overhead remove for get_NN, const for get() methods and some C++11 style for little more performance

File: src/annoylib.h [hunk 6]
```diff
@@ -313,14 +313,13 @@ struct Angular : Base {
   }
   template<typename S, typename T, typename Random>
   static inline void create_split(const vector<Node<S, T>*>& nodes, int f, size_t s, Random& random, Node<S, T>* n) {
-    Node<S, T>* p = (Node<S, T>*)malloc(s); // TODO: avoid
-    Node<S, T>* q = (Node<S, T>*)malloc(s); // TODO: avoid
+    char nodes_buffer[s * 2];
+    Node<S, T>* p = (Node<S, T>*)nodes_buffer;
+    Node<S, T>* q = (Node<S, T>*)nodes_buffer + 1;
```
Comment (erikbern) [inline src/annoylib.h]: this is nice. however i think the second line has to be `    Node<S, T>* q = (Node<S, T>*)(nodes_buffer + s)`

**Comment by erikbern** [comment]
do you mind rebasing on current master?

**Comment by erikbern** [comment]
hm this PR is getting pretty complicated. very interesting features though. but can you break it up into a few separate pull requests?

**Comment by erikbern** [comment]
@ViperCraft there's no way i'm going to be able to merge this at this point – can you please break it up into a series of smaller PRs if you want this to make it into the repo?

**Comment by erikbern** [comment]
@ViperCraft I appreciate this work, but there's no way this can be merged at this point. Will have to be broken up into several smaller PRs. Going to close this for now.

**Pull request closed without merging**